flang/lib/Lower/DirectivesCommon.h

```diff
@@ -808,14 +808,83 @@ genBaseBoundsOps(fir::FirOpBuilder &builder, mlir::Location loc,
   return bounds;
 }
 
+namespace detail {
+template <typename T> //
+static T &&AsRvalueRef(T &&t) {
+  return std::move(t);
+}
+template <typename T> //
+static T AsRvalueRef(T &t) {
+  return t;
+}
+template <typename T> //
+static T AsRvalueRef(const T &t) {
+  return t;
+}
+
+// Helper class for stripping enclosing parentheses and a conversion that
+// preserves type category. This is used for triplet elements, which are
+// always of type integer(kind=8). The lower/upper bounds are converted to
+// an "index" type, which is 64-bit, so the explicit conversion to kind=8
```

Comment:
Ok.

Then I can remove this function to keep the semantics right, but the variable names in some test files will need to be updated to have the explicit "int(..., kind=8)" in there.  Would that be ok?

Reply:
Can you post an example of var name before/after? I guess this is coming from the AsFortran function.

Reply:
One specific difference below.  There are several instances of this with different variables.  The name is in `acc.create`, second line from the end in each snippet.

```
1,9c1,11
<     %28 = fir.convert %27 : (i32) -> index
<     %29 = arith.subi %28, %6 : index
<     %30 = fir.load %3#0 : !fir.ref<i32>
<     %31 = fir.convert %30 : (i32) -> index
<     %32 = arith.subi %31, %6 : index
<     %33 = acc.bounds lowerbound(%29 : index) upperbound(%32 : index) extent(%13 : index) stride(%26#2 : index) startIdx(%6 : index) {strideInBytes = true}
<     %34 = fir.box_addr %15#0 : (!fir.box<!fir.array<?xf32>>) -> !fir.ref<!fir.array<?xf32>>
<     %35 = acc.create varPtr(%34 : !fir.ref<!fir.array<?xf32>>) bounds(%33) -> !fir.ref<!fir.array<?xf32>> {name = "b(n:m)", structured = false}
<     acc.enter_data dataOperands(%35 : !fir.ref<!fir.array<?xf32>>)
---
>     %28 = fir.convert %27 : (i32) -> i64
>     %29 = fir.convert %28 : (i64) -> index
>     %30 = arith.subi %29, %6 : index
>     %31 = fir.load %3#0 : !fir.ref<i32>
>     %32 = fir.convert %31 : (i32) -> i64
>     %33 = fir.convert %32 : (i64) -> index
>     %34 = arith.subi %33, %6 : index
>     %35 = acc.bounds lowerbound(%30 : index) upperbound(%34 : index) extent(%13 : index) stride(%26#2 : index) startIdx(%6 : index) {strideInBytes = true}
>     %36 = fir.box_addr %15#0 : (!fir.box<!fir.array<?xf32>>) -> !fir.ref<!fir.array<?xf32>>
>     %37 = acc.create varPtr(%36 : !fir.ref<!fir.array<?xf32>>) bounds(%35) -> !fir.ref<!fir.array<?xf32>> {name = "b(int(n,kind=8):int(m,kind=8))", structured = false}
>     acc.enter_data dataOperands(%37 : !fir.ref<!fir.array<?xf32>>)
```

Reply:
Ok I see. This is indeed not ideal since the name is meant to be used for error reporting and so on and should at much as possible represent what was in the original code. Would it be possible to peal the expr only when we use it with the asFortran stringstream and keep the real expression to keep the semantic right?

Reply:
Done.

Reply:
Thanks. The changes on the OpenACC and DirectiveCommon side look ok to me. I'll let someone from the OpenMP side to approve the rest of the changes.